lib/services/address/index.js

```diff
@@ -0,0 +1,755 @@
+  $.checkArgument(emitter instanceof EventEmitter, 'First argument is expected to be an EventEmitter');
+  $.checkArgument(Array.isArray(addresses), 'Second argument is expected to be an Array of addresses');
+
+  for(var i = 0; i < addresses.length; i++) {
+    var hashHex = bitcore.Address(addresses[i]).hashBuffer.toString('hex');
+    if(!this.subscriptions[name][hashHex]) {
+      this.subscriptions[name][hashHex] = [];
+    }
+    this.subscriptions[name][hashHex].push(emitter);
+  }
+};
+
+AddressService.prototype.unsubscribe = function(name, emitter, addresses) {
+  $.checkArgument(emitter instanceof EventEmitter, 'First argument is expected to be an EventEmitter');
+  $.checkArgument(Array.isArray(addresses) || _.isUndefined(addresses), 'Second argument is expected to be an Array of addresses or undefined');
+
+  if(!addresses) {
+    return this.unsubscribeAll(name, emitter);
+  }
+
+  for(var i = 0; i < addresses.length; i++) {
+    var hashHex = bitcore.Address(addresses[i]).hashBuffer.toString('hex');
+    if(this.subscriptions[name][hashHex]) {
+      var emitters = this.subscriptions[name][hashHex];
+      var index = emitters.indexOf(emitter);
+      if(index > -1) {
+        emitters.splice(index, 1);
+      }
+    }
+  }
+};
+
+AddressService.prototype.unsubscribeAll = function(name, emitter) {
+  $.checkArgument(emitter instanceof EventEmitter, 'First argument is expected to be an EventEmitter');
+
+  for(var hashHex in this.subscriptions[name]) {
+    var emitters = this.subscriptions[name][hashHex];
+    var index = emitters.indexOf(emitter);
+    if(index > -1) {
+      emitters.splice(index, 1);
+    }
+  }
+};
+
+AddressService.prototype.getBalance = function(address, queryMempool, callback) {
+  this.getUnspentOutputs(address, queryMempool, function(err, outputs) {
+    if(err) {
+      return callback(err);
+    }
+
+    var satoshis = outputs.map(function(output) {
+      return output.satoshis;
+    });
+
+    var sum = satoshis.reduce(function(a, b) {
+      return a + b;
+    }, 0);
+
+    return callback(null, sum);
+  });
+};
+
+/**
+ * @param {String} addressStr - The relevant address
+ * @param {Object} options - Additional options for query the outputs
+ * @param {Number} [options.start] - The relevant start block height
+ * @param {Number} [options.end] - The relevant end block height
+ * @param {Boolean} [options.queryMempool] - Include the mempool in the results
+ * @param {Function} callback
+ */
+AddressService.prototype.getInputs = function(addressStr, options, callback) {
+
+  var self = this;
+
+  var inputs = [];
+  var stream;
+
+  var hashBuffer = bitcore.Address(addressStr).hashBuffer;
+
+  if (options.start && options.end) {
+
+    var endBuffer = new Buffer(4);
+    endBuffer.writeUInt32BE(options.end);
+
+    var startBuffer = new Buffer(4);
+    startBuffer.writeUInt32BE(options.start + 1);
+
+    stream = this.node.services.db.store.createReadStream({
+      gte: Buffer.concat([
```

Comment:
A common pattern in search queries is for the start to be inclusive and the end to be exclusive. So you would do lte: start and gt: end. (We use this pattern in the BitPay codebase.) And here is a supporting reference: http://qedcode.com/content/exclusive-end-dates